proposals/webidl-bindings/idl.md

```diff
@@ -0,0 +1,739 @@
+>readability without requiring complex parsing.
+
+## Wi-IDL Schema {#schema}
+
+The wi-IDL schema language is a language for describing types.
+
+A type is a term that denotes a collection of values. Specifically, there exists
+a unique meta-function mapping values to types (all values have a unique type).
+
+The wi-IDL type schema defines the values that a wi-IDL signature can reference.
+
+There are several different kinds of type:
+
+```
+Type => FunctionType
+Type => QuantifiedType
+Type => TypeInterface
+Type => TupleType
+Type => VariantType
+Type => NominalType
+```
+
+### Function Type {#function-type}
+
+A function type denotes the types of functions. 
+
+```
+FunctionType => TupleType '=>' Type
+```
+
+In the concrete W-Expression form, function types are written:
+
+```
+W-FunctionType ::= '(' 'func' W-TupleType W-Type ')'
+```
+
+For example, the type of a function that returns a string from a pair of
+integers has type -- as an W-Expression term:
+
+```
+(func (integer integer) string)
+```
+
+A function that returns more than one value is indicated by its return type
+being a _TupleType_.
+
+>Discussion Point: webAssembly is described in terms of a stack machine. In that
+>context, all functions simply return a tuple of values. However, in order to
+>support interoperability with systems that are not inherently stack machines,
+>this specification assumes all functions return a single value -- which may be
+>a tuple. In particular, so-called void functions are assumed to return the
+>empty tuple.
+
+A function that may raise an exception can declare it with a throws clause in
+the type:
+
+```
+FunctionType => TupleType '=>' Type 'throws' Type
+```
+or in W-Expression form:
+
+```
+W-FunctionType ::= '(' 'func' W-TupleType W-Type W-Type ')'
+```
+
+The type of an exception is not distinguished from other types. I.e., there is
+no privileged type that is used for exceptions.
+
+### Quantified Type {#quantified-type}
+
+There are two forms of quantifier: universal and existential. 
+
+```
+UniversalType => 'all' TypeVar .. TypeVar '.' Type
+```
+
+and
+
+```
+ExistentialType => 'exist' TypeVar .. TypeVar '.' Type
+```
+
+Or, as W-Expressions:
+
+```
+W-UniversalType ::= '(' 'all' '(' W-TypeVar .. W-TypeVar ')' W-Type ')'
+W-ExistentialType ::= '(' 'exist' '(' W-TypeVar .. W-TypeVar ')' W-Type ')'
+```
+
+Where more than one quantifier is applied to a type, they are left-concatenated; for example:
+
+```
+all X Y . (X Y)
+```
+
+is equivalent to
+
+```
+all X . all Y . (X Y)
+```
+
+There are two forms of TypeVar, corresponding to whether the type variable
+denotes a type or a type function:
+
+```
+TypeVar => Name
+TypeVar => Name '/' Integer
+```
+
+```
+W-TypeVar ::= W-Identifier
+W-TypeVar ::= '(' W-Identifier '/' W-Integer ')'
+```
+
+The second form is used to denote a generic type; for example `Sequence/1`
+denotes the sequence type where the `/1` denotes the fact that `Sequence` takes
+a single type argument.
```

Comment:
That doesn't make much sense to me. If you truly wanted higher-order quantification then you'd need a proper kinding system. A simply arity is not enough, since it doesn't tell whether the respective parameter to the parameter does itself take parameters.